xtensa: use generic cache API

nashif opened the pull request:
use generic cache APIs for intel_adsp and xtensa in general.

nashif committed:
soc: intel_adsp: use abstracted arch APIs for cache ops

Use architecture APIs for cache operations with this soc family.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@lgirdwood  FYI

carlocaione commented on arch/xtensa/core/cache.c:
```diff
@@ -0,0 +1,49 @@
+		z_xtensa_cache_flush(addr, size);
+	} else if (op == K_CACHE_WB_INVD) {
+		z_xtensa_cache_flush_inv(addr, size);
+	} else {
+		return -ENOTSUP;
+	}
+	return 0;
```
useless `return 0` ?

nashif replied:
/home/nashif/zephyrproject/zephyr/arch/xtensa/core/cache.c:34:1: warning: control reaches end of non-void function [-Wreturn-type]
   34 | }
      | ^

carlocaione replied:
Right, nvm.

lyakh replied:
it certainly isn't useless - every `if` branch before the last `else` proceeds to below the whole`if`-`else` so it needs a `return`

carlocaione commented on arch/xtensa/core/cache.c:
```diff
@@ -0,0 +1,49 @@
+		z_xtensa_cache_flush(NULL, Z_DCACHE_MAX);
+	} else if (op == K_CACHE_WB_INVD) {
+		z_xtensa_cache_flush_inv(NULL, Z_DCACHE_MAX);
+	} else {
+		return -ENOTSUP;
+	}
+	return 0;
```
ditto

carlocaione commented on soc/xtensa/intel_adsp/Kconfig.defconfig:
```diff
@@ -19,4 +19,11 @@ config XTENSA_UNCACHED_REGION
+config CACHE_MANAGEMENT
+	def_bool y
+
+choice CACHE_TYPE
+	default HAS_ARCH_CACHE
```
This should be already the default.

carlocaione commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -6,12 +6,14 @@
 #include <adsp_ipc_regs.h>
 #include <adsp_memory.h>
 
+
```
nit: empty space

carlocaione commented on soc/xtensa/intel_adsp/common/multiprocessing.c:
```diff
@@ -12,6 +12,7 @@
 #include <zephyr/sys/__assert.h>
 #include <zephyr/sys/sys_io.h>
 
+
```
ditto.

carlocaione reviewed (dismissed):
Just some small comments.

andyross commented on arch/xtensa/core/cache.c:
```diff
@@ -0,0 +1,49 @@
+	/* not implemented */
+}
+
+int arch_dcache_range(void *addr, size_t size, int op)
```
Can we move this to a header and mark it ALWAYS_INLINE?  The previous API had the advantage of inlining cleanly, so e.g. if you just wanted to flush a pointer to memory of known size and alignment (by far the common case) it would reduce to just one instruction per cache line.  Doing it this way is going to have to invoke the loop code every time, which requires doing a bunch of alignment math on the arguments.

lyakh replied:
Yes, please let's at least allow the compiler eliminate all these branches. In fact I'd actually say the `op` is constant in 100% of the cases, isn't it? When would the operation have to be parametrised and determined at run-time? Don't remember seeing any such cases. In all cases I can think of the caller calls the one and only required operation in that case. So I'm actually not very positive about this API as such...

andyross reviewed (dismissed):
Really not liking the performance regression this implies, though I guess it's possible most of that is tolerable.  The root-ish cause is that IMHO our cache API is pretty clumsy and not amendable to this kind of trickery, but it seems to me that with a little work we could recover the inlining.

andyross commented:
Making the point via analogy instead: imagine this had been our original version, and someone came by to revert this code in favor of using optimized, arch-specific cache APIs for a bunch of drivers, claiming that it saved X hundred bytes of code space and was Y cycles faster at getting DMA posted or whatever.  We'd probably merge that, right?

lgirdwood reviewed (dismissed):
Approve for the generic API side.
@nashif fwiw, I may also need a generic cache line (by address) lock and unlock extension to this API alongside some poisoning support (for later on).
@andyross I'm guessing the (static) inlining would work for you given that the CC can eliminate the ```switch``` in all uses cases in this PR (since the `op`s are all constants) ?

edersondisouza commented on drivers/neural_net/intel_gna.c:
```diff
@@ -460,12 +460,12 @@ static int intel_gna_infer(const struct device *dev,
 	gna_config_desc.labase = (uint32_t)handle->vabase;
 	gna_config_desc.lacnt = (uint16_t)header->layer_count;
-	SOC_DCACHE_FLUSH(&gna_config_desc, sizeof(gna_config_desc));
+	z_xtensa_cache_flush(&gna_config_desc, sizeof(gna_config_desc));
```
Why not remove the macros themselves as well?

nashif replied:
they are still being used outside of Zephyr..

nashif commented:
> Though I don't know that this is all that stable of an API, we have comparatively few cache-aware platforms still. One option might be to take the opportunity to rework the core cache API to more more inlinable/optimizable (basically separate inlines for Icache/Dcache cross flush/invalidate/both _edit: **cross range/all**_), then implement the legacy API as a wrapper.

@carlocaione  ^^

lyakh commented on arch/xtensa/core/cache.c:
```diff
@@ -0,0 +1,49 @@
+
+int arch_dcache_range(void *addr, size_t size, int op)
+{
+	if (op == K_CACHE_INVD) {
```
`switch (op)`?

carlocaione commented:
> > Though I don't know that this is all that stable of an API, we have comparatively few cache-aware platforms still. One option might be to take the opportunity to rework the core cache API to more more inlinable/optimizable (basically separate inlines for Icache/Dcache cross flush/invalidate/both _edit: **cross range/all**_), then implement the legacy API as a wrapper.
> 
> @carlocaione ^^

Alright, I'll try to find time to entirely rework the cache APIs. Stay tuned.

tmleman commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -122,7 +123,7 @@ void power_gate_entry(uint32_t core_id)
 	lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
 	lpsheader->lp_restore_vector = &dsp_restore_vector;
 	soc_cpus_active[core_id] = false;
-	xthal_dcache_all_writeback();
+	sys_cache_data_all(K_CACHE_WB);
```
**_lpsheader_** is in the LPSRAM. Can you confirm that it will work? I have opened an issue on cache for this memory area https://github.com/zephyrproject-rtos/zephyr/issues/49112.

tmleman replied:
@nashif Issue is fixed and xtensa hal function can be replace with _**z_xtensa_cache_flush_inv_all**_. 

I have applied your changes locally and unfortunately they break power gating functionality. Calling _**sys_cache_data_all(K_CACHE_WB)**_ leads to _**z_xtensa_cache_flush(NULL, Z_DCACHE_MAX)**_. It turns out that this function does not flush lpsram cache.

andyross replied:
Wasn't that the bug @edersondisouza fixes a few days ago?

carlocaione commented:
@nashif do you plan to re-work this?

dabekjakub commented on arch/xtensa/core/cache.c:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
```
2023 ;)

dabekjakub reviewed (dismissed):
Looks good to me. I have one question though in the discussion before there was a subject of inlining. Shouldn't the arch_dcache_ functions be inlined too ?

marc-hb commented:
I just submitted 8a9e5fa7ebc1 to SOF CI in https://github.com/thesofproject/sof/pull/6991 and it unfortunately does not compile:
https://github.com/thesofproject/sof/actions/runs/4003607037/jobs/6871863468
```
/zep_workspace/sof/src/idc/zephyr_idc.c: In function 'idc_handler':
/zep_workspace/sof/src/idc/zephyr_idc.c:64:9: error: implicit declaration of function 'SOC_DCACHE_INVALIDATE'; did you mean 'CACHE_INVALIDATE'? [-Werror=implicit-function-declaration]
   64 |         SOC_DCACHE_INVALIDATE(msg, sizeof(*msg));
      |         ^~~~~~~~~~~~~~~~~~~~~
      |         CACHE_INVALIDATE
/zep_workspace/sof/src/idc/zephyr_idc.c: In function 'idc_send_msg':
/zep_workspace/sof/src/idc/zephyr_idc.c:124:17: error: implicit declaration of function 'SOC_DCACHE_FLUSH' [-Werror=implicit-function-declaration]
  124 |                 SOC_DCACHE_FLUSH(payload->data, MIN(sizeof(payload->data), msg->size));
      |                 ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

nashif commented:
> I just submitted [8a9e5fa](https://github.com/zephyrproject-rtos/zephyr/commit/8a9e5fa7ebc136b16ef3758e4c507644927f541f) to SOF CI in [thesofproject/sof#6991](https://github.com/thesofproject/sof/pull/6991) and it unfortunately does not compile: [thesofproject/sof/actions/runs/4003607037/jobs/6871863468](https://github.com/thesofproject/sof/actions/runs/4003607037/jobs/6871863468)

SOF needs to remove the usage of SOC_DCACHE_INVALIDATE and instead use the genric cache api.

marc-hb commented:
> And btw, this PR is failing Zephyr's CI just because of this, so this is known.

Great, would you have a link? I don't see any Github "check" failing right now and I don't see any recent and relevant SOF change either. Or is this internal CI?

nashif commented:
> Great, would you have a link? I don't see any Github "check" failing right now and I don't see any recent and relevant SOF change either. Or is this internal CI?

because it was fixed? 
The fix is here: https://github.com/zephyrproject-rtos/sof/pull/24/files, linked above.

marc-hb commented:
Got it now, thanks! https://github.com/zephyrproject-rtos/sof/commit/5695078f361b is not upstream yet and I wrongly assumed https://github.com/zephyrproject-rtos/sof was never diverging.

marc-hb commented:
BTW https://github.com/zephyrproject-rtos/sof/commit/5695078f361b416c83d7a4f17730ece553de810b also moves the rimage submodule backwards 12 commits. It looks really unrelated, I suspect that wasn't intended. Should be ignored or overridden by `west` most of the time anyway.

marc-hb commented:
After adding the API fix https://github.com/zephyrproject-rtos/sof/commit/5695078f361b416c83d7a4f17730ece553de810b, the firmware now compiles as expected but it stops booting on TGL/ADL: https://github.com/thesofproject/sof/pull/6991

Dunno whether that's expected too or not, just FYI.

nashif committed:
cache: xtensa: implementa arch cache functions

Implement xtensa cache functions supporting the cache API on the
architecture level.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
intel_adsp: move utils to a new header

Move utility code into a new header and cleanup soc.h

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
xtensa: move arch non cache API code from cache.h to arch.h

Move additional cache code related to architecture support into arch.h
and leave cache.h with cache API implementation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
xtensa: use sys_cache API instead of custom interfaces

Use sys_cache instead of custom and internal APIs.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@carlocaione  still having issues with inlining, for example build hello world:

`west build -p -b intel_adsp_cavs25  samples/hello_world/`

you get:
```
b0038590 <parse_manifest>:
{
b0038590:       006136                  entry   a1, 48
 */
__syscall int sys_cache_data_invd_range(void *addr, size_t size);
static inline int z_impl_sys_cache_data_invd_range(void *addr, size_t size)
{
#if defined(CONFIG_CACHE_MANAGEMENT) && defined(CONFIG_DCACHE)
        return cache_data_invd_range(addr, size);
b0038593:       feb3a1                  l32r    a10, b0038060 <_imr_start+0x60> (b0032000 <_heap_sentry+0x11e52000>)
b0038596:       4b3c                    movi.n  a11, 52
b0038598:       0014e5                  call8   b00386e8 <arch_dcache_invd_range>
        for (i = MAN_SKIP_ENTRIES; i < hdr->num_module_entries; i++) {
b003859b:       140c                    movi.n  a4, 1
                union { uintptr_t x; size_t val; } parm1 = { .val = size };
                return (int) arch_syscall_invoke2(parm0.x, parm1.x, K_SYSCALL_SYS_CACHE_DATA_INVD_RANGE);
        }
#endif
        compiler_barrier();
        return z_impl_sys_cache_data_invd_range(addr, size);
```

previously this was inlined correctly without the cache API:

```
b00384fc <parse_manifest>:
{
b00384fc:       004136                  entry   a1, 32
                __asm__ volatile("dhi %0, 0" :: "r"(line));
b00384ff:       fed831                  l32r    a3, b0038060 <_imr_start+0x60> (b0032000 <_heap_sentry+0x11e52000>)
b0038502:       007362                  dhi     a3, 0
b0038505:       fed7a1                  l32r    a10, b0038064 <_imr_start+0x64> (b00320a8 <_heap_sentry+0x11e520a8>)
        for (i = MAN_SKIP_ENTRIES; i < hdr->num_module_entries; i++) {
b0038508:       1e0c                    movi.n  a14, 1
        size_t first = ROUND_DOWN(addr, step);
b003850a:       c0afb2                  movi    a11, -64
b003850d:       002ac6                  j       b00385bc <parse_manifest+0xc0>
        size_t last = ROUND_UP(((long)addr) + bytes, step);
b0038510:       b3a052                  movi    a5, 179
b0038513:       4a5a                    add.n   a4, a10, a5
        size_t first = ROUND_DOWN(addr, step);
b0038515:       102ab0                  and     a2, a10, a11
        size_t last = ROUND_UP(((long)addr) + bytes, step);
b0038518:       1044b0                  and     a4, a4, a11
        for (line = first; bytes && line < last; line += step) {
b003851b:       0001c6                  j       b0038526 <parse_manifest+0x2a>
b003851e:       00                              .byte 00
b003851f:       00                              .byte 00
                __asm__ volatile("dhi %0, 0" :: "r"(line));
b0038520:       007262                  dhi     a2, 0
```
see ` __asm__ volatile("dhi %0, 0" :: "r"(line));`


Looks like we have to many jumps and layering, also invoking system calls that might preventing the functions from being inlined correctly. Not sure how this works for arm64, I looked at generated code and it is similar (or maybe I was building the wrong code / platform). I did test with `west  build -p  -b qemu_cortex_a53_smp tests/kernel/mem_protect/userspace/`

nashif committed:
manifest: update sof revision

Update SOF with latest upstream and cache fixes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
now on top of #57021

nashif committed:
xtensa: cache: add icache implementation from xtensa hal

add icache implementation from xtensa hal.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

marc-hb commented:
SOF and especially @kv2019i has been trying to overcome a fairly large number of regressions and uprev Zephyr in sof/west.yml and I think he's very close now. Could this wait a few more days before merge to reduce the number of balls in the air at the same time? Thanks!

marc-hb commented:
One-line fix submitted in
- #57298
